freud/diffraction.pyx

```diff
@@ -0,0 +1,313 @@
+
+        shift_matrix = np.array(
+            [[1, 0, -roll],
+             [0, 1, -roll],
+             [0, 0, 1]])
+
+        # Translation for [roll_shift, roll_shift]
+        # Then shift using ss
+        shear_matrix = np.array(
+            [[ss[1, 0], ss[0, 0], roll_shift],
+             [ss[1, 1], ss[0, 1], roll_shift],
+             [0, 0, 1]])
+
+        zoom_matrix = np.diag((self.zoom, self.zoom, 1))
+
+        # This matrix uses homogeneous coordinates. It is a 3x3 matrix that
+        # transforms 2D points and adds an offset.
+        inverse_transform = np.linalg.inv(
+            zoom_matrix @ shear_matrix @ shift_matrix)
+
+        img = scipy.ndimage.affine_transform(
+            input=img,
+            matrix=inverse_transform,
+            output_shape=(self.output_size, self.output_size),
+            order=1,
+            mode="constant")
+        return img
+
+    def compute(self, system, view_orientation=None):
+        R"""Computes diffraction pattern.
+
+        Args:
+            system:
+                Any object that is a valid argument to
+                :class:`freud.locality.NeighborQuery.from_system`.
+            view_orientation ((:math:`4`) :class:`numpy.ndarray`, optional):
+                View orientation. Uses :math:`(1, 0, 0, 0)` if not provided
+                or :code:`None` (Default value = :code:`None`).
+        """
+        system = freud.locality.NeighborQuery.from_system(system)
+
+        if view_orientation is None:
+            view_orientation = np.array([1., 0., 0., 0.])
+
+        grid_size = int(self.grid_size / self.zoom)
+
+        # Compute the box projection matrix
+        inv_shear = self._calc_proj(view_orientation, system.box)
+
+        # Rotate points by the view quaternion and shear by the box projection
+        xy = rowan.rotate(view_orientation, system.points)[:, 0:2]
+        xy = xy @ inv_shear.T
+
+        # Map positions to [0, 1] and compute a histogram "image"
+        xy += 0.5
+        xy %= 1
+        im, _, _ = np.histogram2d(
+            xy[:, 0], xy[:, 1], bins=np.linspace(0, 1, grid_size))
+
+        # Compute FFT and convolve with Gaussian
+        cdef double complex[:, :] diffraction_fft
+        diffraction_fft = np.fft.fft2(im)
+        diffraction_fft = scipy.ndimage.fourier.fourier_gaussian(
+            diffraction_fft, self.peak_width / self.zoom)
+        diffraction_fft = np.fft.fftshift(diffraction_fft)
+
+        # Compute the squared modulus of the FFT, which is S(q)
+        self._diffraction = np.real(
+            diffraction_fft * np.conjugate(diffraction_fft))
+
+        # Transform the image (scale, shear, zoom) and normalize S(q) by N^2
+        N = len(system.points)
+        self._diffraction = self._transform(
+            self._diffraction, system.box, inv_shear) / (N*N)
+
+        # Compute a cached array of k-vectors that can be rotated and scaled
+        if not self._called_compute:
```

Comment:
This variable only appears once. I'm guessing it should be first defined in the constructor and then set to `False`. It also needs to be set to `True` at the end of this block where the original kvalues and kvectors are computed.

Reply:
See https://github.com/glotzerlab/freud/blob/58ab65f3a20f001dc9da7b02376f7a9162a012d9/freud/util.pxd

We don't need to do anything differently with this variable. It is owned and managed by the parent class.